crates/re_sdk/src/global.rs

```diff
@@ -28,14 +28,49 @@ impl std::fmt::Display for RecordingScope {
 
 // ---
 
+/// Required to work-around <https://github.com/rerun-io/rerun/issues/2889>
+#[derive(Default)]
+struct ThreadLocalRecording {
+    stream: Option<RecordingStream>,
+}
+
+impl ThreadLocalRecording {
+    fn replace(&mut self, stream: Option<RecordingStream>) -> Option<RecordingStream> {
+        std::mem::replace(&mut self.stream, stream)
+    }
+
+    fn get(&self) -> Option<RecordingStream> {
+        self.stream.clone()
+    }
+}
+
+#[cfg(target_os = "macos")]
+impl Drop for ThreadLocalRecording {
+    fn drop(&mut self) {
+        if let Some(stream) = self.stream.take() {
+            // Work-around for https://github.com/rerun-io/rerun/issues/2889
+            // Calling drop on `self.stream` will panic the calling thread.
+            // But we want to make sure we don't loose the data in the stream.
+            // So how?
+            re_log::warn!("Using thread-local RecordingStream on macOS can result in data loss because of https://github.com/rerun-io/rerun/issues/2889");
+
+            // Give the batchet and sink threads a chance to process the data.
+            std::thread::sleep(std::time::Duration::from_millis(500));
```

Comment:
Tbh not sure if we already talked about this, but there's no way easy way right now to know whether the threads are still busy, right? Would be nice to be able to tell them to join instead 🤔 
Adding 500ms to our shutdown time is pretty awful and error prone.

Reply:
Yeah, it's not great, but we would need to check if all the queues are empty and the threads non-busy all the way down, and it is not trivial

Reply:
I'll open a new issue (and close the old one) as a future TODO

Reply:
https://github.com/rerun-io/rerun/issues/3937